litellm/llms/vertex_ai_and_google_ai_studio/vertex_ai_non_gemini.py

```diff
@@ -243,22 +244,27 @@ def _gemini_convert_messages_with_history(  # noqa: PLR0915
                 contents.append(ContentType(role="model", parts=assistant_content))
 
             ## APPEND TOOL CALL MESSAGES ##
-            if msg_i < len(messages) and (
-                messages[msg_i]["role"] == "tool"
-                or messages[msg_i]["role"] == "function"
-            ):
-                _part = convert_to_gemini_tool_call_result(
-                    messages[msg_i], last_message_with_tool_calls  # type: ignore
-                )
-                contents.append(ContentType(parts=[_part]))  # type: ignore
-                msg_i += 1
+            if msg_i < len(messages):
+                if messages[msg_i]["role"] == "tool" or messages[msg_i]["role"] == "function":
+                    _part = convert_to_gemini_tool_call_result(
+                        messages[msg_i], last_message_with_tool_calls  # type: ignore
+                    )
+                    msg_i += 1
+                    tool_call_responses.append(_part)
+            if msg_i < len(messages):
```

Comment:
why is there a 2nd if block here?

Reply:
The first `if` block groups consecutive `tool_call_response`s into a single `content` by appending them to a list.  The second `if` block adds this combined content to the actual `contents` when a sequence of `tool_call_response`s ends.

This code is placed here, executed immediately after the sequence, to ensure there are no other elements between an assistant(model) response containing `tool_calls` and the corresponding `tool_call_response`s.

The reason for using two separate `if` blocks is that `msg_i` is incremented within the first block.  This separation allows for proper boundary checking of `msg_i` after the increment.